File: golang/3634_min_removal/min_removal.go
// Package main ...
package main

import (
	"sort"
	"testing"

	"github.com/stretchr/testify/assert"
)

func minRemoval(nums []int, k int) int {
	n := len(nums)
	sort.Ints(nums)

	ans := n
	right := 0

	for left := 0; left < n; left++ {
		for right < n && int64(nums[right]) <= int64(nums[left])*int64(k) {
			right++
		}
		current := n - (right - left)
		if current < ans {
			ans = current
		}
	}

	return ans
}

func main() {
	tests := []struct {
		nums []int
		k    int
		ans  int
	}{
		{[]int{2, 1, 5}, 2, 1},
		{[]int{1, 6, 2, 9}, 3, 2},
		{[]int{4, 6}, 2, 0},
	}

	for index, test := range tests {
		assert.Equal(&testing.T{}, test.ans, minRemoval(test.nums, test.k), index)
	}
}

File: cpp/3634_min_removal/min_removal.cpp
#include <algorithm>
#include <cassert>
#include <tuple>
#include <vector>

using namespace std;

class Solution {
 public:
  int minRemoval(vector<int>& nums, int k) {
    int n = nums.size();
    sort(nums.begin(), nums.end());

    int ans = n, right = 0;
    for (int left = 0; left < n; ++left) {
      while (right < n && nums[right] <= static_cast<long long>(nums[left]) * k) {
        ++right;
      }
      ans = min(ans, n - (right - left));
    }

    return ans;
  }
};

int main() {
  vector<tuple<vector<int>, int, int>> tests{
      {{2, 1, 5}, 2, 1},
      {{1, 6, 2, 9}, 3, 2},
      {{4, 6}, 2, 0},
  };

  for (auto [nums, k, ans] : tests) {
    assert(Solution().minRemoval(nums, k) == ans);
  }
}